File: src/minial_agent/api/fs/routers/pptx.py
from fastapi import APIRouter

from minial_agent.api.fs.dependencies import pptx_service
from minial_agent.api.fs.schemas import (
    PptxDeckResponse,
    PptxExportRequest,
    PptxExportResponse,
    PptxOperationRequest,
    PptxOperationResponse,
    PptxSearchResponse,
)


router = APIRouter(prefix="/pptx")


@router.get("/deck", response_model=PptxDeckResponse)
async def get_pptx_deck(user_id: str, uuid: str, path: str) -> PptxDeckResponse:
    return pptx_service.pptx_deck(user_id=user_id, uuid=uuid, path=path)


@router.post("/operations", response_model=PptxOperationResponse)
async def apply_pptx_operations(
    request: PptxOperationRequest,
) -> PptxOperationResponse:
    return pptx_service.pptx_operations(
        user_id=request.user_id,
        uuid=request.uuid,
        path=request.path,
        origin=request.origin,
        expected_revision=request.expected_revision,
        operations=request.operations,
    )


@router.get("/search", response_model=PptxSearchResponse)
async def search_pptx(user_id: str, uuid: str, path: str, q: str, limit: int = 10):
    return pptx_service.pptx_search(
        user_id=user_id,
        uuid=uuid,
        path=path,
        query=q,
        limit=limit,
    )


@router.post("/export/pdf", response_model=PptxExportResponse)
async def export_pptx_pdf(request: PptxExportRequest) -> PptxExportResponse:
    return pptx_service.pptx_export_pdf(
        user_id=request.user_id,
        uuid=request.uuid,
        path=request.path,
    )


@router.post("/export/pptx", response_model=PptxExportResponse)
async def export_pptx_file(request: PptxExportRequest) -> PptxExportResponse:
    return pptx_service.pptx_export_file(
        user_id=request.user_id,
        uuid=request.uuid,
        path=request.path,
    )

File: ui/src/lib/api/pptx.ts
import { apiResourceUrl, apiUrl } from "@/lib/backend-url";

import type {
  PptxDeckResponse,
  PptxExportResponse,
  PptxOperation,
  PptxOperationResponse,
  PptxSearchResponse,
} from "./types";

export async function getPptxDeck({
  userId,
  sessionUuid,
  path,
}: {
  userId: string;
  sessionUuid: string;
  path: string;
}) {
  const params = new URLSearchParams({
    user_id: userId,
    uuid: sessionUuid,
    path,
  });
  const response = await fetch(apiUrl(`/api/fs/pptx/deck?${params.toString()}`));

  if (!response.ok) {
    throw new Error(`PPTX deck load failed: ${response.status}`);
  }

  const deck = (await response.json()) as PptxDeckResponse;
  return {
    ...deck,
    source_url: apiResourceUrl(deck.source_url),
  };
}

export async function applyPptxOperations({
  userId,
  sessionUuid,
  path,
  expectedRevision,
  operations,
}: {
  userId: string;
  sessionUuid: string;
  path: string;
  expectedRevision: number;
  operations: PptxOperation[];
}) {
  const response = await fetch(apiUrl("/api/fs/pptx/operations"), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      user_id: userId,
      uuid: sessionUuid,
      path,
      origin: "user",
      expected_revision: expectedRevision,
      operations,
    }),
  });

  if (!response.ok) {
    throw new Error(`PPTX operation failed: ${response.status}`);
  }

  return (await response.json()) as PptxOperationResponse;
}

export async function searchPptxDeck({
  userId,
  sessionUuid,
  path,
  query,
  limit = 10,
}: {
  userId: string;
  sessionUuid: string;
  path: string;
  query: string;
  limit?: number;
}) {
  const params = new URLSearchParams({
    user_id: userId,
    uuid: sessionUuid,
    path,
    q: query,
    limit: String(limit),
  });
  const response = await fetch(apiUrl(`/api/fs/pptx/search?${params.toString()}`));

  if (!response.ok) {
    throw new Error(`PPTX search failed: ${response.status}`);
  }

  return (await response.json()) as PptxSearchResponse;
}

export async function exportPptx({
  userId,
  sessionUuid,
  path,
  format,
}: {
  userId: string;
  sessionUuid: string;
  path: string;
  format: "pdf" | "pptx";
}) {
  const response = await fetch(apiUrl(`/api/fs/pptx/export/${format}`), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      user_id: userId,
      uuid: sessionUuid,
      path,
    }),
  });

  if (!response.ok) {
    throw new Error(`PPTX export failed: ${response.status}`);
  }

  return (await response.json()) as PptxExportResponse;
}
